DOC: Add documentation explaining our promotion rules

seberg opened the pull request:
This adds a dedicated page about promotion rules.

---

Please don't hesitate to rip it apart, I haven't tried to really iterate yet, but I think it would be useful to get feedback.  (Also don't hesitate to push changes/fixes!)

My thought was to try and not dwell too long on the technical details, but start with the things that I thought are more interesting to very new users.  That might mean the plot (taken from the NEP, but without the text), may be a little bit underexplained.

There is also the question if more examples would be relevant, although I don't really want to overdo it (in the NEP the big table is important, for a new user with fresh eyes, I am not quite sure what is.)

I need to figure out if we have any docs that need to be removed/refer here (beyond `result_type, `promote_types`, and maybe the release note/migration guide as a link).

*Doctest will probably fail for now, and references might be bad... let's fix it as we iterate*

---

@mdhaber maybe you can have a look?

seberg committed:
DOC: Add documentation explaining our promotion rules

This adds a dedicated page about promotion rules.

jakevdp commented on doc/source/reference/figures/nep-0050-promotion-no-fonts.svg:
Maybe this is too detailed, but one potential confusion in this diagram is that in some cases multiple promotion paths are available. For example, with `int8 + uint8`, it's unclear from the diagram alone whether the result would be `int16` or `float16`.

seberg replied:
Ah, right: I think in the NEP I added one sentence that the lowest precision and kind is chosen (i.e. the most to the left and down) that is higher or equal all involved.
That seems reasonable to me at least, but the sentence below the image needs adjusting to point it out.

mdhaber replied:
> the lowest precision and kind is chosen (i.e. the most to the left and down...)

I'm still not sure that is unambiguous. Perhaps it can be made unambiguous by thinking of each level of kind and precision as adding additional weight (or distance) to edges of the graph.

Then

"Input dtypes are promoted to the node with kind and precision at least as high as those of the inputs which minimizes the total distance traveled along the graph."

Alternatively,

"The result dtype is the one that minimizes distance traveled along the graph under the constraint that its kind and precision are at least as high as those of the original dtypes."

Or:
```
min   distance traveled
s.t.  result kind ≥ max(input kind)
      result precision ≥ max(input precision)
```

If this is accurate, I'd suggest:
- Adding arrows to the edges to indicate allowed movement
- Using color or stroke thickness to indicate weight of edges

seberg replied:
Maybe note that normally you never cross kind boundary because one of the inputs will have the same kind?
I doubt it matters much for that same reason, but happy to try to specify more: unnecessarily crossing the kind won't seem like a minimal solution even if you could give "kind" almost 0 weight or so and then it works.

mdhaber replied:
I'm trying to make this part precise, though. Please either suggest a precise statement that you think is easier to understand than one of the above, or LMK which of the formulations about you prefer.

seberg replied:
How about syaing that you pick the lowest precision which is higher or equal than the inputs while using the highest input kind (with few exceptions where a higher precision can only be achieved by using a higher kind)?

I think the emphasis should be on precision, because the result kind is one of the input kinds with the only *strange* exception of `uint + int -> float` sometimes.

mdhaber replied:
As described, the "precision" of `uint64` and `int64` are the same (64). `int` is the higher input `kind` of the two. So `int64` would seem to satisfy this criterion as-stated.

I think you are assuming the reader to know that the effective precision of `uint` is higher than the effective precision of `int` with the same precision number, but that is not documented.

seberg replied:
Hmmm, in the plot I put the precision of the unsigned versions to be higher than the signed versions to make that work...
(uint64 covers more numbers than int64 if you consider the negative numbers due to the higher "kind".  Just like float32 has less precison than int32 due to the smaller mantissa)

mdhaber replied:
> in the plot I put the precision of the unsigned versions to be higher than the signed versions

OK, but this is out of sync with the text, in which the precision is just the number appended to the kind.

Is there a technical way to distinguish between these two meanings of "precision"?

> float32 has less precison than int32 due to the smaller mantissa

So why does `float64` have higher precision than `int64`?

seberg replied:
> there a technical way to distinguish between these two meanings of "precision"?

I don't know, I don't have a better idea.  We don't have to use "precision" at all.  You could say range, but that is even more confusing.

> So why does float64 have higher precision than int64?

To make sense for promotion...  If you prefer add arrows that point in the wrong direction.  Either way, the NumPy promotion/cast-safety rules have always been inconsistent here, so you can't really draw a fully consistent picture.

mdhaber replied:
OK, so whatever it's called, it just means further right on the diagram. I'll see if I can work with that.

mdhaber replied:
> The input dtype with the higher kind determines the kind of the result dtype. The result dtype has precision as low as possible without appearing to the left of either input dtype in the diagram. 

> There is one exception to the first rule: because there are no dtypes of kind `int` that appear to the right of `uint64`, the results of operations between `uint64` and any `int` are promoted to `float64`.

seberg replied:
Sounds good.

seberg replied:
Adopted that, just the first sentence, because the exception is already below. (makes things a bit shorter).

seberg committed:
Try clarifying the promotion figure a bit more and maybe fix doc warnings/errors

mdhaber committed:
DOC: adjustments to promotion rule documentation

seberg committed:
Small edits

mhvk commented on doc/source/reference/arrays.promotion.rst:
```diff
@@ -0,0 +1,263 @@
+  np.int8(-56)
+  ... RuntimeWarning: overflow encountered in scalar add
+
+Note that NumPy warns when overflows occur for scalars, but not for arrays;
```
Hmm, that's actually surprising in itself, since we were thinking of making array scalars and arrays more similar...

mhvk commented on doc/source/reference/arrays.promotion.rst:
```diff
@@ -0,0 +1,263 @@
+3. A special case the above leads to is that NumPy will promote many
+   combinations of signed and unsigned integers to ``float64`` because no
+   integer dtype can hold both inputs.
+   This can unfortunately be 
```
Incomplete sentence

seberg replied:
Just removed it, I think this is already a lot for a niche detail. (slight tweak above)

mhvk commented on doc/source/reference/arrays.promotion.rst:
```diff
@@ -0,0 +1,263 @@
+
+The precision here comes from the bit size of the numerical value but
+an ``int32`` cannot always be stored in a ``float32`` without loss of
+precision, this leads to the 
```
And another one. I think both can be removed, since this basically has been covered.

mhvk replied:
Don't forget to delete these lines (l164-166) - this is already covered (and incomplete here).

seberg replied:
Ouch, sorry, missed this!

mhvk commented on doc/source/reference/arrays.promotion.rst:
```diff
@@ -0,0 +1,263 @@
+The reason for this is that integer summations are otherwise very likely
+to overflow and give confusing results.
+This rule also applies to the underlying ``np.add.reduce``, ``np.multiply.reduce``
+and other reduction methods.
```
Does it apply to all reductions over `int`, i.e., also `np.minimum.reduce` and `np.maximum.reduce`?

seberg replied:
no, those exact ones are the only special ones right now.

mhvk replied:
Then it would seem best to delete "and other reduction methods".

seberg replied:
Deleted.  Maybe I thought of mean, or so... But those go to double anyway.

mhvk commented on doc/source/reference/arrays.promotion.rst:
```diff
@@ -0,0 +1,263 @@
+  between any NumPy integer with any Python integer value by casting both
+  to ``float64``.
+
+It should be noted that while these exceptions can apply to more functions,
```
I would phrase this in a friendly way as,
```
In principle, some of these exceptions may make sense for other functions.
Please raise an issue if you feel this is the case.
```
(eg., really `np.clip` is another case where we may want exceptions...)

seberg replied:
adopted

mhvk commented on doc/source/reference/arrays.promotion.rst:
```diff
@@ -0,0 +1,263 @@
+
+The following rules apply:
+* NumPy byte strings (``np.bytes_``) can be promoted to unicode strings
+  (``np.str_``).  A ``result_type``
```
Incomplete! What does it do for non-ascii?

seberg replied:
I don't why non-ascii is relevant, since you are promoting to unicode.  Added that the length of both is taken into account.

mhvk commented on doc/source/reference/arrays.promotion.rst:
```diff
@@ -0,0 +1,263 @@
+    Some of these rules describe the state as of NumPy 2.0 and details
+    may be surprising and are under consideration to be change in the future.
+    However, changes always have to be weighed against backwards compatibility
+    concerns.
```
Add: "Please raise an issue if you have particular ideas about how promotion should work"

mhvk commented on doc/source/reference/arrays.promotion.rst:
```diff
@@ -0,0 +1,263 @@
+
+
+.. [#hist-reasons]: To a large degree, this may just be for choices made early
+   on in NumPy's predecessors even.  You may find some more details also in
```
Second sentence: For more details, see `NEP...`

seberg replied:
adopted

mhvk reviewed (commented):
This looks good! Some smaller comments in-line.

I think one further one is to explain that the decision has been to generally *not* care about values of anything, but only about their data type; the comparisons are then an exception to that rule for python scalars.

mattip commented:
ping @seberg

seberg commented:
@mdhaber did you want to make another pass?  Otherwise, maybe I'll just address Marten's smaller comments and we can push it through for now.

seberg committed:
Address Martens review comments

seberg committed:
Also adopt a suggestion by Matt (slightly rephrasing anothe part)

seberg commented:
Since we didn't iterate further, I addressed Marten's changes and adopted the one suggestion (the first bit, so that should be sanity checked).  Also rebased, last two commits are new.

Maybe we should just put this in, and maybe @mdhaber can just do a PR if interested in polishing?

mhvk commented on doc/source/reference/arrays.promotion.rst:
```diff
@@ -0,0 +1,262 @@
+   example, the result of an operation involving ``int64`` and ``float16``
+   is ``float64``.
+2. When mixing unsigned and signed integers with the same precision, the
+   result will has *higher* precision than either inputs. Additionally,
```
will has -> will have

mhvk commented on doc/source/reference/arrays.promotion.rst:
```diff
@@ -0,0 +1,262 @@
+  >> np.int16(1) + 1.0
+  np.float64(2.0)
+
+However, these rules can also lead to surprising behaviors when working with
```
behaviors -> behavior (or suprising behaviors -> surprises)

mhvk commented on doc/source/reference/arrays.promotion.rst:
```diff
@@ -0,0 +1,262 @@
+Numerical promotion
+-------------------
+
+The following images shows the numerical promotion rules with the kinds
```
images -> image

mhvk commented on doc/source/reference/arrays.promotion.rst:
```diff
@@ -0,0 +1,262 @@
+precision, this leads to the 
+
+
+Notable or special promotion behaviors
```
behaviors -> behavior (cannot be plural)

Maybe, though, better as "Exceptions to the general promotion rules"?

seberg replied:
Yeah, let's go with that.

mhvk commented on doc/source/reference/arrays.promotion.rst:
```diff
@@ -0,0 +1,262 @@
+but the operation will sometimes dictate that result.
+Division always returns floating point values and comparison always booleans.
+
+This leads to some special behaviors which may appear as "exceptions" to
```
behaviors -> behavior. But perhaps best to shorten to "This leads to what may appear as "exceptions" to the rules"

mhvk commented on doc/source/reference/arrays.promotion.rst:
```diff
@@ -0,0 +1,262 @@
+    However, changes always have to be weighed against backwards compatibility
+    concerns.
+    Please raise an issue if you have particular ideas about how promotion
+    should work
```
Missing period at end of sentence.

mhvk commented on doc/source/reference/arrays.promotion.rst:
```diff
@@ -0,0 +1,262 @@
+
+The following rules apply:
+* NumPy byte strings (``np.bytes_``) can be promoted to unicode strings
+  (``np.str_``).  The promotion takes into account the length of both.
```
My confusion here is how promotion works if the bytes are not proper unicode. Does one get an error?

seberg replied:
Ah, no I see.  You are thinking of utf-8, I'll mention UCS1 (latin1) encoding.

seberg replied:
Ah, nvm. We are indeed erroring on non-ascii in the cast for clarity.

seberg replied:
Replacing the second sentence with:

> However, casting the bytes to unicode will fail for non-ascii characters.

mhvk commented on doc/source/reference/arrays.promotion.rst:
```diff
@@ -0,0 +1,262 @@
+* NumPy ``timedelta`` can in some cases promote with integers.
+
+.. note::
+    Some of these rules describe the state as of NumPy 2.0 and details
```
Maybe
```
Some of these rules are somewhat surprising, and are being considered for
change in the future. However, any backward-incompatible changes have to
be weighed against the risks of breaking existing code. Please raise an issue
if you have particular ideas about how promotion should work.
```

seberg replied:
Reads well, adopted, thanks.

mhvk commented on doc/source/reference/arrays.promotion.rst:
```diff
@@ -0,0 +1,262 @@
+
+Details of promoted ``dtype`` instances
+---------------------------------------
+The above promotion mainly refers to the behavior when mixing different DType
```
Maybe: "The above discussion has mainly dealt with the behavior ..."

mhvk commented on doc/source/reference/arrays.promotion.rst:
```diff
@@ -0,0 +1,262 @@
+* Resulting structured dtypes will be packed (but aligned if inputs were).
+
+This behaviors is the best behavior for most programs where storage details
+are not relevant to the final results, but the use of incorrect byte-order
```
but the use -> and where the use (otherwise it feels like it is a problem to use native byte order; at least, I read it that wrong way the first time).

mhvk commented on doc/source/reference/arrays.promotion.rst:
```diff
@@ -0,0 +1,262 @@
+
+
+.. [#hist-reasons]: To a large degree, this may just be for choices made early
+   on in NumPy's predecessors even.  For more details, see `NEP 50 <NEP50>`.
```
delete "even"

mhvk commented on doc/source/reference/arrays.promotion.rst:
```diff
@@ -0,0 +1,262 @@
+   on in NumPy's predecessors even.  For more details, see `NEP 50 <NEP50>`.
+
+.. [#NEP50]: See also `NEP 50 <NEP50>` which changed the rules for NumPy 2.0.
+   previous versions of NumPy would sometimes return higher precision results
```
Capitalize "previous"

mhvk reviewed (commented):
@seberg - some final nitpicks on the promotion docs.

seberg committed:
Address/adopt Martens review comments

seberg commented:
Thanks for the quick and thorough check, @mhvk!

mhvk reviewed (approved):
Great, I think we should indeed get this in now.

seberg commented:
OK, I guess let's put it in then, thanks everyone.
